src/Compatibility/Microsoft.DotNet.ApiCompatibility/Rules/AssemblyIdentityMustMatch.cs

```diff
@@ -0,0 +1,98 @@
+            string leftAssemblyCulture = left.Identity.CultureName == string.Empty ? "neutral" : left.Identity.CultureName;
+            Version leftAssemblyVersion = left.Identity.Version;
+            ReadOnlySpan<byte> leftAssemblyPublicKeyToken = left.Identity.PublicKeyToken.AsSpan();
+
+            string rightAssemblyName = right.Identity.Name;
+            string rightAssemblyCulture = right.Identity.CultureName == string.Empty ? "neutral" : right.Identity.CultureName;
+            Version rightAssemblyVersion = right.Identity.Version;
+            ReadOnlySpan<byte> rightAssemblyPublicKeyToken = right.Identity.PublicKeyToken.AsSpan();
+
+            if (leftAssemblyName != rightAssemblyName)
+            {
+                differences.Add(new CompatDifference(DiagnosticIds.AssemblyIdentityMustMatch, string.Format(Resources.AssemblyNameDoesNotMatch, leftAssemblyName, rightAssemblyName, leftName, rightName), DifferenceType.Changed, right.Identity.GetDisplayName()));
+            }
+
+            if (leftAssemblyCulture != rightAssemblyCulture)
+            {
+                differences.Add(new CompatDifference(DiagnosticIds.AssemblyIdentityMustMatch, string.Format(Resources.AssembyCultureDoesNotMatch, leftAssemblyCulture, rightAssemblyCulture, leftName, rightName), DifferenceType.Changed, right.Identity.GetDisplayName()));
+            }
+ 
+            if (rightAssemblyVersion < leftAssemblyVersion)
+            {
+                differences.Add(new CompatDifference(DiagnosticIds.AssemblyIdentityMustMatch, string.Format(Resources.AssembyVersionDoesNotMatch, rightAssemblyVersion.ToString(), leftAssemblyVersion.ToString(), rightName, leftName), DifferenceType.Changed, right.Identity.GetDisplayName()));
+            }
+            else if (Settings.StrictMode && leftAssemblyVersion < rightAssemblyVersion)
+            {
+                differences.Add(new CompatDifference(DiagnosticIds.AssemblyIdentityMustMatch, string.Format(Resources.AssembyVersionDoesNotMatch, leftAssemblyVersion.ToString(), rightAssemblyVersion.ToString(), leftName, rightName), DifferenceType.Changed, right.Identity.GetDisplayName()));
+            }
+
+            bool isLeftRetargetable = (left.Identity.Flags & AssemblyNameFlags.Retargetable) != 0;
+            bool isRightRetargetable = (right.Identity.Flags & AssemblyNameFlags.Retargetable) != 0;
+            if (!leftAssemblyPublicKeyToken.IsEmpty && !isLeftRetargetable && !leftAssemblyPublicKeyToken.SequenceEqual(rightAssemblyPublicKeyToken))
+            {
+                differences.Add(
+                    new CompatDifference(
+                        DiagnosticIds.AssemblyIdentityMustMatch,
+                        string.Format(Resources.AssemblyPublicKeyTokenDoesNotMatch, GetStringRepresentation(leftAssemblyPublicKeyToken), GetStringRepresentation(rightAssemblyPublicKeyToken), leftName, rightName),
+                        DifferenceType.Changed,
```

Comment:
since the differenceType and diagnostic ID is the same across all of these differences, you could add a helper method that takes the message and the symbol/target in, then return the compat difference. Share that method and just call it everywhere to get a new compat difference. Something like: https://github.com/dotnet/sdk/blob/56a6239ba8959df4e7d851923e5022a0fce78805/src/Compatibility/Microsoft.DotNet.ApiCompatibility/Rules/MembersMustExist.cs#L70